### PR TITLE
Bias improvement #15

### PR DESCRIPTION
**Pull request by TiagoMAntunes**
Bias term is now being accepted in MOELinear.apply as requested in #15 

The solution is 2 dimensional and each block will be handling 32 columns. Lines don't matter: only one block will be called vertically, and this block will sum all the following ones. Thanks @laekov for the suggestion

I added different tensor types for the numerical stability tests. I had to set a different precision for the half tensors.

Let me know if anything isn't clear

### CODE REVIEW

File: cuda/moe.cpp [hunk 1]
```diff
@@ -33,36 +33,54 @@ std::vector<torch::Tensor> moe_local_gather(
...
+	
+	input_buf = at::cat({input_buf, ones}, 1); // [X 1]
+}
+
 std::vector<torch::Tensor> moe_forward(
-        torch::Tensor input_buf,     // [batch_size x in_feat]
-        torch::Tensor weight,        // [num_expert x out_feat x in_feat]
-        torch::Tensor expert_count   // [batch_size]
+        torch::Tensor input_buf,     		// [batch_size x in_feat]
+		torch::Tensor expert_count,  		// [batch_size]
```
Comment (laekov) [inline cuda/moe.cpp:52]: Seems like you are mixing up tab and spaces. Let's expand all tabs to 4 spaces.

**Comment by TiagoMAntunes** [comment]
I've changed the previous changes into CUDA. I developed a reduction kernel that is currently performing the job and will work on improving it at a later time this week, which will only consist on changing the kernel itself and nothing else.

**Comment by TiagoMAntunes** [comment]
I also fixed the indentation in `moe_compute_kernel.cu` with 4 spaces as asked by @laekov 
Let's try and keep consistent with them now!

**Comment by zjujh1995** [comment]
This PR leaded to a CUDA Error during training as follows:
`cuda/moe_compute_kernel.cu:223 code=13(CUBLAS_STATUS_EXECUTION_FAILED) "cublasXgemm( smgr->handle(i), CUBLAS_OP_N, CUBLAS_OP_N, in_feat, expert_count[i], out_feat, &alpha, weight + i * in_feat * out_feat, in_feat, grad_output_buf + ptr * out_feat, out_feat, &beta, grad_input_buf + in_feat * ptr, in_feat )"`
Everything goes fine before these modification. Pls make a check.
Looking forward to receiving your reply. Thanks a lot!

**Comment by TiagoMAntunes** [comment]
> This PR leaded to a CUDA Error during training as follows:
> `cuda/moe_compute_kernel.cu:223 code=13(CUBLAS_STATUS_EXECUTION_FAILED) "cublasXgemm( smgr->handle(i), CUBLAS_OP_N, CUBLAS_OP_N, in_feat, expert_count[i], out_feat, &alpha, weight + i * in_feat * out_feat, in_feat, grad_output_buf + ptr * out_feat, out_feat, &beta, grad_input_buf + in_feat * ptr, in_feat )"`
> Everything goes fine before these modification. Pls make a check.
> Looking forward to receiving your reply. Thanks a lot!
> 
> PS: The bug only appears in FP16 model with APEX.

Hi @zjujh1995 thanks for the feedback. I am quite busy at the moment, so it would be easier for me if you could describe more accurately how to replicate the error or provide a base script, this way I can quickly look into it. Thanks!

**Comment by zjujh1995** [comment]
> > This PR leaded to a CUDA Error during training as follows:
> > `cuda/moe_compute_kernel.cu:223 code=13(CUBLAS_STATUS_EXECUTION_FAILED) "cublasXgemm( smgr->handle(i), CUBLAS_OP_N, CUBLAS_OP_N, in_feat, expert_count[i], out_feat, &alpha, weight + i * in_feat * out_feat, in_feat, grad_output_buf + ptr * out_feat, out_feat, &beta, grad_input_buf + in_feat * ptr, in_feat )"`
> > Everything goes fine before these modification. Pls make a check.
> > Looking forward to receiving your reply. Thanks a lot!
> > PS: The bug only appears in FP16 model with APEX.
> 
> Hi @zjujh1995 thanks for the feedback. I am quite busy at the moment, so it would be easier for me if you could describe more accurately how to replicate the error or provide a base script, this way I can quickly look into it. Thanks!

Thanks for your reply. We didn't modify anything but only use the official codes of Megatrom-LM (e.g. running pretrain_gpt.sh). As a result, when we fmoefied the GPT model and set fp16=True, the model crashed as we described above.
Grateful for your effort & looking forward to solving the case.

**Comment by zjujh1995** [comment]
> > This PR leaded to a CUDA Error during training as follows:
> > `cuda/moe_compute_kernel.cu:223 code=13(CUBLAS_STATUS_EXECUTION_FAILED) "cublasXgemm( smgr->handle(i), CUBLAS_OP_N, CUBLAS_OP_N, in_feat, expert_count[i], out_feat, &alpha, weight + i * in_feat * out_feat, in_feat, grad_output_buf + ptr * out_feat, out_feat, &beta, grad_input_buf + in_feat * ptr, in_feat )"`
> > Everything goes fine before these modification. Pls make a check.
> > Looking forward to receiving your reply. Thanks a lot!
> > PS: The bug only appears in FP16 model with APEX.
> 
> Hi @zjujh1995 thanks for the feedback. I am quite busy at the moment, so it would be easier for me if you could describe more accurately how to replicate the error or provide a base script, this way I can quickly look into it. Thanks!

Please don't hesitate to inform us if you make any further progress on the problem. Many thanks!

**Comment by TiagoMAntunes** [comment]
> > > This PR leaded to a CUDA Error during training as follows:
> > > `cuda/moe_compute_kernel.cu:223 code=13(CUBLAS_STATUS_EXECUTION_FAILED) "cublasXgemm( smgr->handle(i), CUBLAS_OP_N, CUBLAS_OP_N, in_feat, expert_count[i], out_feat, &alpha, weight + i * in_feat * out_feat, in_feat, grad_output_buf + ptr * out_feat, out_feat, &beta, grad_input_buf + in_feat * ptr, in_feat )"`
> > > Everything goes fine before these modification. Pls make a check.
> > > Looking forward to receiving your reply. Thanks a lot!
> > > PS: The bug only appears in FP16 model with APEX.
> > 
> > 
> > Hi @zjujh1995 thanks for the feedback. I am quite busy at the moment, so it would be easier for me if you could describe more accurately how to replicate the error or provide a base script, this way I can quickly look into it. Thanks!
> 
> Please don't hesitate to inform us if you make any further progress on the problem. Many thanks!
> 
> Update: We've found if we change `scalar_t` of line 249 in moe_compute_kernel.cu into `float`, fmoefied models will run correctly in both fp16 and fp32.
> 
> BTW, may I ask why you implement the bias operation manually instead of using cuBLAS's beta (just as describing in Issue `[Roadmap] v0.2 release plan`) ? Will the manual implement hurt calculation efficiency?

It is using cuBLAS beta parameter. Check moe_compute_kernel.cu:166 it's the forward function
The new kernel that's causing the crash is in the backwards function. Since we're not using PyTorch layers, it requires manual implementation for autograd, therefore the `column_reduce` kernel.

Thank you for your feedback. I have some urgents things to solve in short-term but I will try to look into this in the following days

**Comment by TiagoMAntunes** [comment]
@zjujh1995 Hi, could you try running it again? I ran my Megatron now with this fix and it worked, but I was getting code=14 instead of code=13 as you specified.

**Comment by TiagoMAntunes** [comment]
This is ready to be merged. Let me know if any errors are found

**Comment by zjujh1995** [comment]
> @zjujh1995 Hi, could you try running it again? I ran my Megatron now with this fix and it worked, but I was getting code=14 instead of code=13 as you specified.

Thanks for your effort. The problem disappeared with the help of your recent commits.

**Comment by laekov** [comment]
nicely done

**Review by laekov — APPROVED** [review]

**Pull request merged**